### PR TITLE
Integration with peer2peer networking

### PR DESCRIPTION
**Pull request by coot**
This PR integrates with peer2peer networking and allows to run a node in either mode:

* non-p2p: the only officially supported version
* p2p: unverified and unsupported p2p mode

This also includes:

* a new topology file format for p2p nodes
* a mechanism to reload topology file on SIGHUP signal

- network trace instances: use ToJSON or ToObject instances
- network trace instances: peer selection traces
- network trace instances: ToJSON instances
- consensus trace instances: trace termination reason
- P2P / NonP2P cardano-node
- Update p2p topology configuration via the SIGHUP signal
- Removed MonoLocalBinds extension
- topology file: improved parsing error message
- Improve NetworkTopology generator
- byron-mainnet configuration: added p2p trace options
- p2p topology file for cardano mainnet
- Document the p2p topology file
- Updated cardano-node-chairman
- Updated trace-dispatcher and trace-forward libraries
- Updated tx-generator
- Updated cardano-testnet package
- Code cleanup in Tracing.Config
- nixos service: add p2p topology support.
- nixos-service: Add space-cost profiling support
- nixos-service: Add space-heap profiling support

### CODE REVIEW

**Comment by coot** [comment]
Although `p2p-master` in `ouroboros-network` hasn't been merged yet (and thus the git dependencies check failure), this is ready for review.

File: cardano-node/src/Cardano/Tracing/Tracers.hs [hunk 1]
```diff
@@ -623,7 +691,9 @@ mkConsensusTracers mbEKGDirect trSel verb tr nodeKern fStats = do
    traceServedCount Nothing _ = pure ()
    traceServedCount (Just ekgDirect) ev =
      when (isRollForward ev) $
-       sendEKGDirectCounter ekgDirect "cardano.node.metrics.served.header.counter.int"
+       sendEKGDirectCounter ekgDirect
+                            "cardano.node.metrics.served.header.counter.int"
+
```
Comment (karknu) [inline cardano-node/src/Cardano/Tracing/Tracers.hs:696]: This change seems unnecessary.

File: cardano-node/src/Cardano/Tracing/Tracers.hs [hunk 2]
```diff
@@ -565,6 +633,7 @@ mkConsensusTracers mbEKGDirect trSel verb tr nodeKern fStats = do
             tracerOnOff (traceBlockFetchClient trSel) verb "BlockFetchClient" tr
     , Consensus.blockFetchServerTracer = traceBlockFetchServerMetrics trmet meta tBlocksServed
         tLocalUp tMaxSlotNo $ tracerOnOff (traceBlockFetchServer trSel) verb "BlockFetchServer" tr
+    , Consensus.keepAliveClientTracer = tracerOnOff (traceKeepAliveClient trSel) verb "KeepAliveClient" tr
```
Comment (karknu) [inline cardano-node/src/Cardano/Tracing/Tracers.hs:635]: The movement of this line doesn't seem needed.

File: cardano-node/src/Cardano/Tracing/Tracers.hs [hunk 3]
```diff
@@ -678,8 +748,8 @@ cdfCounter v !size !step tCdf= do
     when (v < lim) $
         STM.modifyTVar' tCdf (\(CdfCounter c) -> CdfCounter $ c + step)
 
-    CdfCounter cdf <- STM.readTVar tCdf
-    return $! fromIntegral cdf / fromIntegral size
+    (CdfCounter cdf) <- STM.readTVar tCdf
+    return $! (fromIntegral cdf / fromIntegral size)
```
Comment (karknu) [inline cardano-node/src/Cardano/Tracing/Tracers.hs:751]: This is just readding some unneeded '('.

File: cardano-node/src/Cardano/Tracing/Tracers.hs [hunk 4]
```diff
@@ -1136,11 +1206,14 @@ nodeToClientTracers'
...
+      tracerOnOff (traceLocalChainSyncProtocol trSel)
+                  verb "LocalChainSyncProtocol" tr
   , NodeToClient.tTxSubmissionTracer =
-    tracerOnOff (traceLocalTxSubmissionProtocol trSel) verb "LocalTxSubmissionProtocol" tr
+      tracerOnOff (traceLocalTxSubmissionProtocol trSel)
+                  verb "LocalTxSubmissionProtocol" tr
   , NodeToClient.tStateQueryTracer =
-    tracerOnOff (traceLocalStateQueryProtocol trSel) verb "LocalStateQueryProtocol" tr
+      tracerOnOff (traceLocalStateQueryProtocol trSel)
+                  verb "LocalStateQueryProtocol" tr
```
Comment (karknu) [inline cardano-node/src/Cardano/Tracing/Tracers.hs:1215]: Not needed.

**Review by karknu — APPROVED** [review]
LGTM.
Just some minor comments.

File: cardano-node/src/Cardano/Tracing/Tracers.hs (hunk 1 above)
Comment (coot) [reply cardano-node/src/Cardano/Tracing/Tracers.hs:696]: It fits the line into 80 columns.

File: cardano-node/src/Cardano/Tracing/Tracers.hs (hunk 2 above)
Comment (coot) [reply cardano-node/src/Cardano/Tracing/Tracers.hs:635]: The reason for that was to keep all node-to-node mini-protocol tracers together.

File: cardano-node/src/Cardano/Tracing/Tracers.hs (hunk 4 above)
Comment (coot) [reply cardano-node/src/Cardano/Tracing/Tracers.hs:1215]: this splits too long lines.  I'll put these changes into a separate commit.

File: cardano-node/src/Cardano/Node/Configuration/TopologyP2P.hs [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+                                 <*> (o .: "PublicRoots"                                    )
+                                 <*> (o .:? "useLedgerAfterSlot" .!= UseLedger DontUseLedger)
+
+instance ToJSON NetworkTopology where
+  toJSON top =
+    case top of
+      RealNodeTopology lrpg prp ul -> object [ "LocalRoots"         .= lrpg
+                                             , "PublicRoots"        .= prp
+                                             , "useLedgerAfterSlot" .= ul
+                                             ]
```
Comment (newhoggy) [inline cardano-node/src/Cardano/Node/Configuration/TopologyP2P.hs:172]: What's happening here where sometimes we use `PublicRoots` and sometimes `publicRoots`?

File: doc/getting-started/understanding-config-files.md [hunk 6]
```diff
@@ -21,6 +21,64 @@ Tells your node to which nodes in the network it should talk to. A minimal versi
...
+    resolved ips.
+
+* `valency` tells the node how many connections your node should try to pick from the given group. If a dns address is given, valency governs to how many resolved ip addresses should we maintain active (hot) connection.
+
+Your __block-producing__ node must __ONLY__ talk to your __relay nodes__, and the relay node should talk to other relay nodes in the network.
+
+You __can__ tell the node that the topology configuration file changed by sending a SIGHUP
+signal to the `cardano-node` process, e.g. `pkill -HUP cardano-node`. After receiving the
+signal, `cardano-node` will re-read the file and restart all dns resolution. Please
+**note** that this only applies to the topology configuration file!
```
Comment (newhoggy) [inline doc/getting-started/understanding-config-files.md:81]: Does this work or is there an equivalent for Windows?

File: nix/nixos/cardano-node-service.nix [hunk 7]
```diff
@@ -331,21 +363,62 @@ in {
...
-          Static routes to peers, specific to a given instance (when multiple instances are used).
+          Static routes to local peers, specific to a given instance (when multiple instances are used).
+        '';
+      };
+
+      useNewTopology = mkOption {
+        type = types.bool;
+        default = cfg.nodeConfig.EnableP2P or false;
+        description = ''
+          Use new, p2p/ledger peers combatible topology.
```
Comment (newhoggy) [inline nix/nixos/cardano-node-service.nix:410]: `combatible` typo?

**Review by newhoggy — APPROVED** [review]
Approved with comments.

File: doc/getting-started/understanding-config-files.md (hunk 6 above)
Comment (coot) [reply doc/getting-started/understanding-config-files.md:81]: Is `*nix` only, and unlikely needed on Windows.

File: nix/nixos/cardano-node-service.nix (hunk 7 above)
Comment (coot) [reply nix/nixos/cardano-node-service.nix:410]: I'll fix it.

File: cardano-node/src/Cardano/Node/Configuration/TopologyP2P.hs (hunk 5 above)
Comment (coot) [reply cardano-node/src/Cardano/Node/Configuration/TopologyP2P.hs:172]: The difference becomes easier to grasp with an example (which is pushed to documentation):

```json
{
  "LocalRoots": {
    "groups": [
      {
        "localRoots": {
          "addrs": [
            {
              "addr": "x.x.x.x",
              "port": 3001
            }
          ],
          "advertise": false
        },
        "valency": 1
      }
    ]
  },
  "PublicRoots": [
    {
      "publicRoots" : {
        "addrs": [
          {
            "addr": "y.y.y.y",
            "port": 3002
          }
        ],
        "advertise": false
      }
    }
  ],
  "useLedgerAfterSlot": 0
}
```
`PublicRoots` is the top level record name, which has various fields, including `publicRoots`; the same for `LocalRoots` and its field `localRoots`.

File: cardano-node/src/Cardano/Tracing/OrphanInstances/Network.hs [hunk 8]
```diff
@@ -827,61 +854,79 @@ instance ToObject (AnyMessageAndAgency ps)
...
+        Aeson.object [ "address" .= toJSON ip
+                     , "port" .= show port
+                     ]
+    toJSON (SockAddrInet6 port _ addr _) =
+        let ip = IP.fromHostAddress6 addr in
+        Aeson.object [ "address" .= toJSON ip
+                     , "port" .= show port
+                     ]
+    toJSON (SockAddrUnix path) =
+        Aeson.object [ "path" .= show path ]
```
Comment (Jimbo4350) [inline cardano-node/src/Cardano/Tracing/OrphanInstances/Network.hs:1152]: "socketPath" would be a better name

File: cardano-node/src/Cardano/Tracing/OrphanInstances/Network.hs [hunk 9]
```diff
@@ -924,12 +970,13 @@ instance ToObject (WithDomainName (SubscriptionTrace SockAddr)) where
              , "event" .= show ev ]
 
 
-instance (Show peer) => ToObject (WithMuxBearer peer MuxTrace) where
-  toObject _verb (WithMuxBearer b ev) =
+instance (ToObject peer) => ToObject (WithMuxBearer peer MuxTrace) where
```
Comment (Jimbo4350) [inline cardano-node/src/Cardano/Tracing/OrphanInstances/Network.hs:1278]: Redundant brackets, this may cause CI to fail

File: cardano-node/src/Cardano/Tracing/OrphanInstances/Network.hs [hunk 10]
```diff
@@ -1351,6 +1507,45 @@ instance ToObject (TracePeerSelection SockAddr) where
     mkObject [ "kind" .= String "ChurnWait"
              , "diffTime" .= toJSON dt
              ]
+  toObject _verb (TraceChurnMode c) =
+    mkObject [ "kind" .= String "ChurnMode"
+             , "event" .= show c ]
+
+instance Aeson.ToJSON AbstractState where
```
Comment (Jimbo4350) [inline cardano-node/src/Cardano/Tracing/OrphanInstances/Network.hs:1514]: Comments on what each of the constructors in `AbstractState` means would be useful

File: cardano-node/src/Cardano/Node/Run.hs [hunk 11]
```diff
@@ -420,6 +435,48 @@ handleSimpleNode scp runP p2pMode trace nodeTracers nc onKernel = do
     forM_ basicInfoItems $ \(LogObject nm mt content) ->
       traceNamedObject (appendName nm tr) (mt, content)
 
+#ifdef UNIX
+  updateVars :: LOMeta
```
Comment (Jimbo4350) [inline cardano-node/src/Cardano/Node/Run.hs:439]: A comment and perhaps a better name for this function would be good. `updateVars` is very generic what about `nodeTopologyUpdate` or something similar?

File: cardano-node/src/Cardano/Tracing/OrphanInstances/Network.hs (hunk 8 above)
Comment (coot) [reply cardano-node/src/Cardano/Tracing/OrphanInstances/Network.hs:1152]: Done

File: cardano-node/src/Cardano/Tracing/OrphanInstances/Network.hs (hunk 9 above)
Comment (coot) [reply cardano-node/src/Cardano/Tracing/OrphanInstances/Network.hs:1278]: done

File: doc/getting-started/understanding-config-files.md [hunk 12]
```diff
@@ -21,6 +21,64 @@ Tells your node to which nodes in the network it should talk to. A minimal versi
...
+        ],
+        "advertise": false
+      }
+    }
+  ],
+  "useLedgerAfterSlot": 0
+}
+```
+
+* The main difference between `LocalRoots` and `PublicRoots` is that with the former
```
Comment (Jimbo4350) [inline doc/getting-started/understanding-config-files.md:63]: Why not "LocalPeers" and "PublicPeers"? Roots seems a bit opaque to me.

File: cardano-testnet/src/Testnet/Byron.hs [hunk 13]
```diff
@@ -69,22 +78,59 @@ defaultTestnetOptions = TestnetOptions
...
 replaceNodeLog :: Int -> String -> String
 replaceNodeLog n s = T.unpack (T.replace "logs/node-0.log" replacement (T.pack s))
   where replacement = T.pack ("logs/node-" <> show @Int n <> ".log")
 
 -- | Rewrite a line in the configuration file
-rewriteConfiguration :: Int -> String -> String
-rewriteConfiguration _ "TraceBlockchainTime: False" = "TraceBlockchainTime: True"
-rewriteConfiguration n s | "logs/node-0.log" `L.isInfixOf` s = replaceNodeLog n s
-rewriteConfiguration _ s = s
+rewriteConfiguration :: Bool -> Int -> String -> String
```
Comment (Jimbo4350) [inline cardano-testnet/src/Testnet/Byron.hs:89]: I'm just seeing this function for the first time. This seems extremely hacky but looks like something to be fixed in another PR.

File: cardano-testnet/src/Testnet/Byron.hs [hunk 14]
```diff
@@ -69,22 +78,59 @@ defaultTestnetOptions = TestnetOptions
...
+rewriteConfiguration True _ "EnableP2P: False"                 = "EnableP2P: True"
+rewriteConfiguration False _ "EnableP2P: True"                 = "EnableP2P: False"
+rewriteConfiguration _ _ s                                     = s
 
 rewriteParams :: TestnetOptions -> Value -> Value
 rewriteParams testnetOptions = rewriteObject
   $ HM.insert "slotDuration" (J.toJSON @String (show @Int (slotDuration testnetOptions)))
 
+mkTopologyConfig :: Int -> Int -> [Int] -> Bool -> ByteString
+mkTopologyConfig i numBftNodes allPorts False = J.encode topologyNonP2P
```
Comment (Jimbo4350) [inline cardano-testnet/src/Testnet/Byron.hs:101]: Boolean blindness.

File: cardano-testnet/src/Testnet/Shelley.hs [hunk 15]
```diff
@@ -116,6 +131,41 @@ rewriteGenesisSpec testnetOptions startTime =
 startTimeOffsetSeconds :: DTC.NominalDiffTime
 startTimeOffsetSeconds = if OS.isWin32 then 90 else 15
 
+
+mkTopologyConfig :: Int -> [Int] -> Int -> Bool -> ByteString
```
Comment (Jimbo4350) [inline cardano-testnet/src/Testnet/Shelley.hs:135]: This function has been duplicated, why?

**Review by Jimbo4350 — CHANGES_REQUESTED** [review]
Looking good but a few comments

File: cardano-node/src/Cardano/Tracing/OrphanInstances/Network.hs [hunk 16]
```diff
@@ -941,6 +1036,55 @@ instance Show remotePeer => ToObject (TraceKeepAliveClient remotePeer) where
...
+  toObject _verb (FetchingNewLedgerState cnt) =
+    mkObject
+      [ "kind" .= String "FetchingNewLedgerState"
+      , "numberOfPools" .= cnt
+      ]
+  toObject _verb DisabledLedgerPeers =
+    mkObject
+      [ "kind" .= String "DisabledLedgerPeers"
+      ]
+  toObject _verb (TraceUseLedgerAfter ula) =
```
[the remark this answers is not in the document]
Comment (coot) [reply cardano-node/src/Cardano/Tracing/OrphanInstances/Network.hs:1223]: The p2p node can take root peers from the ledger.  Depending how it is configured it will either trace that this is disabled or enabled (after some slot).

File: cabal.project [hunk 17]
```diff
@@ -235,8 +235,8 @@ source-repository-package
 source-repository-package
   type: git
   location: https://github.com/input-output-hk/ouroboros-network
-  tag: 94782e5ca52f234ff8eeddc6322a46cca0b69c0e
-  --sha256: 1da3pka4pn6sjf6w19d957aryjc9ip1a3g0vz7jz66pjri3v2n0j
+  tag: 82032848afaf348b7f06a5b591abae58f03d5aa3
```
[the remark this answers is not in the document]
Comment (coot) [reply cabal.project]: No, I am just about to update this.

File: cardano-node/src/Cardano/Node/Configuration/POM.hs [hunk 18]
```diff
@@ -38,8 +45,29 @@ import           Ouroboros.Consensus.Mempool.API (MempoolCapacityBytesOverride (
...
-
-import qualified Data.Aeson.Types as Aeson
+import qualified Ouroboros.Consensus.Node as Consensus ( NetworkP2PMode (..) )
+
+data NetworkP2PMode = EnabledP2PMode | DisabledP2PMode
+  deriving (Eq, Show, Generic)
+
+data SomeNetworkP2PMode where
+    SomeNetworkP2PMode :: forall p2p.
+                          Consensus.NetworkP2PMode p2p
```
[the remark this answers is not in the document]
Comment (coot) [reply cardano-node/src/Cardano/Node/Configuration/POM.hs:55]: diffusion is parametrised with a type level `Diffusion.P2P` kind.  We also need to represent this in a config file, where the existential wrapper is useful.

File: doc/getting-started/understanding-config-files.md (hunk 12 above)
Comment (coot) [reply doc/getting-started/understanding-config-files.md:63]: These are not just peers, these are indeed the roots of the network graph.  That's consistent with our naming convention in the code base (and documentation, though that's mostly haddock right now).

File: cardano-testnet/src/Testnet/Shelley.hs (hunk 15 above)
Comment (coot) [reply cardano-testnet/src/Testnet/Shelley.hs:135]: I am not sure right now (the PR was a team effort), I'll look into it.

File: cardano-node/src/Cardano/Node/Configuration/TopologyP2P.hs [hunk 19]
```diff
@@ -0,0 +1,201 @@
...
+instance FromJSON UseLedger where
+  parseJSON (Data.Aeson.Number n) =
+    if n >= 0 then return $ UseLedger $ UseLedgerAfter $ SlotNo $ floor n
+              else return $ UseLedger   DontUseLedger
+  parseJSON _ = mzero
+
+instance ToJSON UseLedger where
+  toJSON (UseLedger (UseLedgerAfter (SlotNo n))) = Number $ fromIntegral n
+  toJSON (UseLedger DontUseLedger)               = Number (-1)
+
```
[the remark this answers is not in the document]
Comment (karknu) [reply cardano-node/src/Cardano/Node/Configuration/TopologyP2P.hs:54]: No, this is one option in the p2p capable topology file.

File: cardano-node/src/Cardano/Tracing/OrphanInstances/Network.hs [hunk 20]
```diff
@@ -941,6 +1036,55 @@ instance Show remotePeer => ToObject (TraceKeepAliveClient remotePeer) where
...
+      , "addresses" .= show addrs
+      ]
+  toObject _verb (FetchingNewLedgerState cnt) =
+    mkObject
+      [ "kind" .= String "FetchingNewLedgerState"
+      , "numberOfPools" .= cnt
+      ]
+  toObject _verb DisabledLedgerPeers =
+    mkObject
+      [ "kind" .= String "DisabledLedgerPeers"
```
[the remark this answers is not in the document]
Comment (coot) [reply cardano-node/src/Cardano/Tracing/OrphanInstances/Network.hs:1221]: explained below

File: cardano-node/src/Cardano/Node/Configuration/POM.hs [hunk 21]
```diff
@@ -358,6 +449,28 @@ makeNodeConfiguration pnc = do
   diffusionMode <- lastToEither "Missing DiffusionMode" $ pncDiffusionMode pnc
   snapshotInterval <- lastToEither "Missing SnapshotInterval" $ pncSnapshotInterval pnc
 
+  ncTargetNumberOfRootPeers <-
```
[the remark this answers is not in the document]
Comment (coot) [reply cardano-node/src/Cardano/Node/Configuration/POM.hs:452]: Remove them from `NodeConfiguration`? Then how they would be available in `Cardano.Node.Run`?

They must be provided for p2p node (`PartialNodeConfiguration` provides the default values), they are not used by non p2p node.

File: cardano-node/src/Cardano/Node/Run.hs (hunk 11 above)
Comment (coot) [reply cardano-node/src/Cardano/Node/Run.hs:439]: I went for `updateTopologyConfiguration`

File: cardano-testnet/src/Testnet/Shelley.hs (hunk 15 above)
Comment (coot) [reply cardano-testnet/src/Testnet/Shelley.hs:135]: It's not a duplicated function, it has a different type signature and the meaning of its arguments is slightly different.  It's probably cleaner to leave this as is and not generalise further.

File: cardano-testnet/src/Testnet/Byron.hs (hunk 14 above)
Comment (coot) [reply cardano-testnet/src/Testnet/Byron.hs:101]: I added haddocks.

File: cardano-node/src/Cardano/Node/Configuration/Socket.hs [hunk 22]
```diff
@@ -150,7 +152,7 @@ gatherConfiguredSockets NodeConfiguration { ncNodeIPv4Addr,
       (Nothing, Just sock)  -> return (Just (ActualSocket sock))
       (Just _, Just _)      -> throwError ClashingPublicIpv6SocketGiven
       (Just addr, Nothing)  ->
-              fmap SocketInfo . head
+              fmap (SocketInfo . addrAddress) . head
```
[the remark this answers is not in the document]
Comment (coot) [reply cardano-node/src/Cardano/Node/Configuration/Socket.hs:155]: yes, I agree we should show throw a proper error if one requested an address that is not available on the host.

File: cardano-node/src/Cardano/Node/Configuration/TopologyP2P.hs [hunk 23]
```diff
@@ -0,0 +1,201 @@
...
+import           Cardano.Slotting.Slot (SlotNo (..))
+
+import           Cardano.Node.Types
+import           Cardano.Node.Configuration.Topology (TopologyError (..))
+
+import           Ouroboros.Network.NodeToNode (PeerAdvertise (..))
+import           Ouroboros.Network.PeerSelection.LedgerPeers (UseLedgerAfter (..))
+import           Ouroboros.Network.PeerSelection.RelayAccessPoint (RelayAccessPoint (..))
+
+newtype UseLedger = UseLedger UseLedgerAfter deriving (Eq, Show)
```
[the remark this answers is not in the document]
Comment (coot) [reply cardano-node/src/Cardano/Node/Configuration/TopologyP2P.hs:43]: done

File: cardano-node/src/Cardano/Node/Configuration/TopologyP2P.hs [hunk 24]
```diff
@@ -0,0 +1,201 @@
...
+instance ToJSON NodeSetup where
+  toJSON ns =
+    object
+      [ "nodeId"             .= nodeId ns
+      , "nodeIPv4Address"    .= nodeIPv4Address ns
+      , "nodeIPv6Address"    .= nodeIPv6Address ns
+      , "producers"          .= producers ns
+      , "useLedgerAfterSlot" .= useLedger ns
+      ]
+
```
[the remark this answers is not in the document]
Comment (coot) [reply cardano-node/src/Cardano/Node/Configuration/TopologyP2P.hs:81]: @karknu what do you think about renaming this to something like `RootConfig`, it's more than just and address.

File: cardano-node/src/Cardano/Node/Configuration/TopologyP2P.hs [hunk 25]
```diff
@@ -0,0 +1,201 @@
...
+instance ToJSON RootAddress where
+  toJSON ra =
+    object
+      [ "addrs" .= addrs ra
+      , "advertise" .= advertise ra
+      ]
+
+-- | Transforms a 'RootAddress' into a pair of 'RelayAccessPoint' and its
+-- corresponding 'PeerAdvertise' value.
```
[the remark this answers is not in the document]
Comment (coot) [reply cardano-node/src/Cardano/Node/Configuration/TopologyP2P.hs:101]: done

File: cardano-node/src/Cardano/Node/Configuration/TopologyP2P.hs [hunk 26]
```diff
@@ -0,0 +1,201 @@
...
+--
+rootAddressToRelayAccessPoint
+  :: RootAddress
+  -> [(RelayAccessPoint, PeerAdvertise)]
+rootAddressToRelayAccessPoint RootAddress { addrs, advertise } =
+    [ (ra, advertise) | ra <- addrs ]
+
+data LocalRootPeers = LocalRootPeers
```
[the remark this answers is not in the document]
Comment (coot) [reply cardano-node/src/Cardano/Node/Configuration/TopologyP2P.hs:109]: @karknu I think we should rename this type to `LocalRootPeerGroup`, as this configures a single local root group.  This would also fit with the other name `LocalRootPeersGroups ~ [LocalRootPeerGroup]`.  Do you agree?

File: cardano-node/src/Cardano/Node/Configuration/TopologyP2P.hs [hunk 27]
```diff
@@ -0,0 +1,201 @@
...
+  toJSON top =
+    case top of
+      RealNodeTopology lrpg prp ul -> object [ "LocalRoots"         .= lrpg
+                                             , "PublicRoots"        .= prp
+                                             , "useLedgerAfterSlot" .= ul
+                                             ]
+
+-- | Read the `NetworkTopology` configuration from the specified file.
+--
+readTopologyFile :: NodeConfiguration -> IO (Either Text NetworkTopology)
```
[the remark this answers is not in the document]
Comment (coot) [reply cardano-node/src/Cardano/Node/Configuration/TopologyP2P.hs:176]: That's true, I don't mind it, because eventually non-p2p stuff will be removed.

File: cardano-node/src/Cardano/Node/Run.hs [hunk 28]
```diff
@@ -194,28 +216,37 @@ handlePeersListSimple tr nodeKern = forever $ do
...
   -> (NodeKernel IO RemoteConnectionId LocalConnectionId blk -> IO ())
   -- ^ Called on the 'NodeKernel' after creating it, but before the network
   -- layer is initialised.  This implies this function must not block,
   -- otherwise the node won't actually start.
   -> IO ()
-handleSimpleNode scp runP trace nodeTracers nc onKernel = do
+handleSimpleNode scp runP p2pMode trace nodeTracers nc onKernel = do
   meta <- mkLOMeta Notice Public
 
+  traceNamedObject
```
[the remark this answers is not in the document]
Comment (coot) [reply cardano-node/src/Cardano/Node/Run.hs:249]: yes, it should!

File: cardano-node/src/Cardano/Node/Run.hs [hunk 29]
```diff
@@ -387,76 +468,85 @@ checkVRFFilePermissions vrfPrivKey = do
...
-    -- TODO: these limits are arbitrary at the moment;
-    -- issue: https://github.com/input-output-hk/ouroboros-network/issues/1836
-    , daAcceptedConnectionsLimit = AcceptedConnectionsLimit {
-        acceptedConnectionsHardLimit = 512
-      , acceptedConnectionsSoftLimit = 384
-      , acceptedConnectionsDelay     = 5
+
+mkP2PArguments
+  :: NodeConfiguration
+  -> STM IO [(Int, Map RelayAccessPoint PeerAdvertise)]
```
[the remark this answers is not in the document]
Comment (coot) [reply cardano-node/src/Cardano/Node/Run.hs:531]: Configuration of local root peers, the same as `LocalRootPeersGroups` but presented as diffusion takes it.  I'd be keen to provide `LocalRootPeersGroups` or something similar in `ouroboros-network`, but that's outside of this pr.

File: cardano-node/src/Cardano/Tracing/OrphanInstances/Network.hs (hunk 10 above)
Comment (coot) [reply cardano-node/src/Cardano/Tracing/OrphanInstances/Network.hs:1514]: That's connection manager's abstract state, useful for debugging.  We have a whole chapter about it, and this isn't the right place to provide documentation.  I think a link to the spec should be enough.

Comment (coot) [reply cardano-node/src/Cardano/Tracing/OrphanInstances/Network.hs:1514]: done

File: cabal.project (hunk 17 above)
[the remark this answers is not in the document]
Comment (coot) [reply cabal.project]: done, both `ouroboros-network` and `ekg-forward` branches are merged.

File: cardano-node/src/Cardano/Node/Configuration/POM.hs (hunk 21 above)
[the remark this answers is not in the document]
Comment (Jimbo4350) [reply cardano-node/src/Cardano/Node/Configuration/POM.hs:452]: I just read your other comment that we will eventually be removing the non-p2p stuff so this is fine,

File: cardano-node/src/Cardano/Node/Configuration/TopologyP2P.hs (hunk 19 above)
[the remark this answers is not in the document]
Comment (Jimbo4350) [reply cardano-node/src/Cardano/Node/Configuration/TopologyP2P.hs:61]: There are now two definitions for `NodeSetup` which is confusing. Renaming the other can work as well.

File: cardano-node/src/Cardano/Node/Configuration/TopologyP2P.hs (hunk 25 above)
[the remark this answers is not in the document]
Comment (Jimbo4350) [reply cardano-node/src/Cardano/Node/Configuration/TopologyP2P.hs:116]: :+1: Where?

File: cardano-node/src/Cardano/Node/Run.hs (hunk 29 above)
[the remark this answers is not in the document]
Comment (Jimbo4350) [reply cardano-node/src/Cardano/Node/Run.hs:532]: :+1: can you leave a comment explaining what the `Int` refers to?

File: cardano-node/src/Cardano/Tracing/OrphanInstances/Network.hs (hunk 20 above)
[the remark this answers is not in the document]
Comment (Jimbo4350) [reply cardano-node/src/Cardano/Tracing/OrphanInstances/Network.hs:1221]: Where?

File: doc/getting-started/understanding-config-files.md (hunk 12 above)
Comment (Jimbo4350) [reply doc/getting-started/understanding-config-files.md:63]: Can we just clarify that root essentially means initial?

File: cardano-node/src/Cardano/Tracing/OrphanInstances/Network.hs (hunk 20 above)
[the remark this answers is not in the document]
Comment (coot) [reply cardano-node/src/Cardano/Tracing/OrphanInstances/Network.hs:1221]: https://github.com/input-output-hk/cardano-node/pull/3363/files#r749666294

File: cardano-node/src/Cardano/Node/Configuration/TopologyP2P.hs (hunk 25 above)
[the remark this answers is not in the document]
Comment (coot) [reply cardano-node/src/Cardano/Node/Configuration/TopologyP2P.hs:116]: Check the `RootConfig` type where both `RelayAccessPoint` and `PeerAdvertise` are introduced in this module.

File: doc/getting-started/understanding-config-files.md (hunk 12 above)
Comment (coot) [reply doc/getting-started/understanding-config-files.md:63]: Yes, I'll do it.

File: cardano-node/src/Cardano/Tracing/OrphanInstances/Network.hs (hunk 20 above)
[the remark this answers is not in the document]
Comment (Jimbo4350) [reply cardano-node/src/Cardano/Tracing/OrphanInstances/Network.hs:1224]: Right, can we add a comment?

File: cardano-node/src/Cardano/Node/Configuration/TopologyP2P.hs (hunk 25 above)
[the remark this answers is not in the document]
Comment (Jimbo4350) [reply cardano-node/src/Cardano/Node/Configuration/TopologyP2P.hs:116]: :+1: thanks

**Review by Jimbo4350 — APPROVED** [review]

File: cardano-node/src/Cardano/Node/Run.hs (hunk 29 above)
[the remark this answers is not in the document]
Comment (coot) [reply cardano-node/src/Cardano/Node/Run.hs:532]: done

**Comment by coot** [comment]
bors merge

**Review by denisshevchenko — APPROVED** [review]

**Comment by coot** [comment]
bors merge
